File: alert/webhook.go
package alert

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/Alexmaster12345/IntrusionShield/detection"
)

type WebhookNotifier struct {
	url string
}

type webhookPayload struct {
	Timestamp string `json:"timestamp"`
	Msg       string `json:"msg"`
	Severity  int    `json:"severity"`
	Protocol  string `json:"protocol"`
	SrcIP     string `json:"src_ip"`
	DstIP     string `json:"dst_ip"`
	SrcPort   uint16 `json:"src_port"`
	DstPort   uint16 `json:"dst_port"`
}

func (w *WebhookNotifier) Send(a detection.Alert) error {
	payload := webhookPayload{
		Timestamp: a.Timestamp.Format(time.RFC3339),
		Msg:       a.Msg,
		Severity:  a.Severity,
		Protocol:  a.Protocol,
		SrcPort:   a.SrcPort,
		DstPort:   a.DstPort,
	}
	if a.SrcIP != nil {
		payload.SrcIP = a.SrcIP.String()
	}
	if a.DstIP != nil {
		payload.DstIP = a.DstIP.String()
	}

	body, _ := json.Marshal(payload)
	return w.post(body)
}

func (w *WebhookNotifier) SendRaw(msg string) error {
	body, _ := json.Marshal(map[string]string{"message": msg})
	return w.post(body)
}

func (w *WebhookNotifier) post(body []byte) error {
	resp, err := http.Post(w.url, "application/json", bytes.NewReader(body))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return fmt.Errorf("webhook returned %d", resp.StatusCode)
	}
	return nil
}

File: alert/webhook.py
import json
import urllib.request
from datetime import datetime

from detection.engine import Alert


class WebhookNotifier:
    def __init__(self, url: str):
        self.url = url

    def send(self, a: Alert) -> None:
        payload = {
            "timestamp": a.timestamp.isoformat(),
            "msg": a.msg,
            "severity": a.severity,
            "protocol": a.protocol,
            "src_ip": a.src_ip,
            "dst_ip": a.dst_ip,
            "src_port": a.src_port,
            "dst_port": a.dst_port,
        }
        self._post(payload)

    def send_raw(self, msg: str) -> None:
        self._post({"message": msg})

    def _post(self, payload: dict) -> None:
        body = json.dumps(payload).encode()
        req = urllib.request.Request(
            self.url, data=body,
            headers={"Content-Type": "application/json"},
        )
        with urllib.request.urlopen(req, timeout=5) as resp:
            if resp.status >= 300:
                raise RuntimeError(f"Webhook returned {resp.status}")
